File: lintcode/easy/remove_duplicates_from_sorted_list/py/remove_duplicates_from_sorted_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-duplicates-from-sorted-list
@Language: Python
@Datetime: 17-02-15 23:49
'''

"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: A ListNode
    @return: A ListNode
    """
    def deleteDuplicates(self, head):
        if head == None:
            return None
        
        slow = head
        fast = head.next
        
        while fast != None:
            if fast.val != slow.val:
                slow.next = fast
                slow      = slow.next
                
            fast = fast.next
            
        slow.next = None
        
        return head

File: lintcode/easy/remove_duplicates_from_sorted_list/cpp/remove_duplicates_from_sorted_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/remove-duplicates-from-sorted-list
@Language: C++
@Datetime: 17-02-15 23:51
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: head node
     */
    ListNode *deleteDuplicates(ListNode *head) 
    {
        if (head == nullptr) {
            return nullptr;
        }
        
        auto slow = head;
        auto fast = head->next;
        
        while (fast != nullptr) {
            if (slow->val != fast->val) {
                slow->next = fast;
                slow       = slow->next;
            }
            fast = fast->next;
        }
        
        slow->next = nullptr;
        
        return head;
    }
};